File: src/renderer/src/types/Structure.ts
/**
 * Interfaces and types for structure data
 */
export interface IStructureData {
	materials: IMaterialData[]
	sections: ISectionData[]
	nodes: INodeData[]
	bars: IBarData[]
	supports: ISupportData[]
	loads: ILoadData[]
	results: IResultsData[]
}

// Material and Section ///////////////////////////////////////////////////////////////////////////
export interface IMaterialData {
	name: string
	properties: { E: number; G: number; ni: number; rho: number }
}

export interface ISectionData {
	name: string
	area: number
	inertias: { Ix: number; Iy: number; Iz: number }
}

// Nodes //////////////////////////////////////////////////////////////////////////////////////////
export interface INodeData {
	name: string
	position: [number, number, number]
	type: 'Fixed' | 'Hinged' | 'spring' | 'blend'
}

// bars ///////////////////////////////////////////////////////////////////////////////////////////
export interface IBarData {
	name: string
	start_node: string
	end_node: string
	section: string
	material: string
	rotation: number
	releases: IReleasesData[]
}

// Releases ***************************************************************************************
export type IReleasesData =
	| 'Dxi'
	| 'Dyi'
	| 'Dzi'
	| 'Rxi'
	| 'Ryi'
	| 'Rzi'
	| 'Dxj'
	| 'Dyj'
	| 'Dzj'
	| 'Rxj'
	| 'Ryj'
	| 'Rzj'

// Supports ///////////////////////////////////////////////////////////////////////////////////////
export interface ISupportData {
	node: string
	supports: {
		Dx: number | boolean
		Dy: number | boolean
		Dz: number | boolean
		Rx: number | boolean
		Ry: number | boolean
		Rz: number | boolean
	}
}

// Loads //////////////////////////////////////////////////////////////////////////////////////////
export interface ILoadData {
	name: string
	nodes: IPointLoadsData[]
	bars: IBarLoadsData
}

// Nodal loads ************************************************************************************
export interface IPointLoadsData {
	name: string
	node: string
	loads: { Fx: number; Fy: number; Fz: number; Mx: number; My: number; Mz: number }
}

// Bar loads **************************************************************************************
export interface IBarLoadsData {
	point: IBarPointLoadsData[]
	distributed: IBarDistributedLoadsData[]
}

export interface IBarPointLoadsData {
	name: string
	bar: string
	position: number
	system: 'local' | 'global'
	loads: { Fx: number; Fy: number; Fz: number; Mx: number; My: number; Mz: number }
}

export interface IBarDistributedLoadsData {
	name: string
	bar: string
	position: [number, number] // start, end
	system: 'local' | 'global'
	loads: {
		Fx: [number, number]
		Fy: [number, number]
		Fz: [number, number]
		Mx: [number, number]
		My: [number, number]
		Mz: [number, number]
	}
}

export type forcesType = 'Fx' | 'Fy' | 'Fz' | 'Mx' | 'My' | 'Mz'

// Results ////////////////////////////////////////////////////////////////////////////////////////
export interface IResultsData {
	load_case: string
	displacements: IDisplacementResultsData[]
	reactions: IReactionsData[]
	extreme_forces: IExtremeForcesData[]
}

// Displacements **********************************************************************************
export interface IDisplacementResultsData {
	node: string
	Dx: number
	Dy: number
	Dz: number
	Rx: number
	Ry: number
	Rz: number
}

// Reactions **************************************************************************************
export interface IReactionsData {
	node: string
	Fx: number
	Fy: number
	Fz: number
	Mx: number
	My: number
	Mz: number
}

// Extreme Forces *********************************************************************************
export interface IExtremeForcesData {
	bar: string
	// Start
	Fxi: number
	Fyi: number
	Fzi: number
	Mxi: number
	Myi: number
	Mzi: number

	// End
	Fxj: number
	Fyj: number
	Fzj: number
	Mxj: number
	Myj: number
	Mzj: number
}

File: src/backend/pyengineer/types/structure.py
"""Typing for structure"""
from typing import Literal

from pydantic import BaseModel

# Material Interface //////////////////////////////////////////////////////////////////////////////
class IMaterialProperties(BaseModel):
    """Interface for material properties"""
    E: float
    G: float
    nu: float
    rho: float

class IMaterial(BaseModel):
    """Interface for material"""
    name: str
    properties: IMaterialProperties

# Section Interface ///////////////////////////////////////////////////////////////////////////////
class ISectionInertias(BaseModel):
    """Interface for inertias"""
    Ix: float
    Iy: float
    Iz: float

class ISection(BaseModel):
    """Interface for section"""
    name: str
    area: float
    inertias: ISectionInertias

# Node Interface //////////////////////////////////////////////////////////////////////////////////
class INode(BaseModel):
    """Interface for nodes"""
    name: str
    position: list[float]  # [x, y, z]

# Bar Interface ///////////////////////////////////////////////////////////////////////////////////
class IBar(BaseModel):
    """Interface for bars"""
    name: str
    start_node: str
    end_node: str
    section: str
    material: str
    rotation: float
    releases: list[Literal['Dxi', 'Dyi', 'Dzi',
                           'Rxi', 'Ryi', 'Rzi',
                           'Dxj', 'Dyj', 'Dzj',
                           'Rxj', 'Ryj', 'Rzj']]

# Support Interface ///////////////////////////////////////////////////////////////////////////////
class ISupportType(BaseModel):
    """Interface for support types"""
    Dx: bool | float
    Dy: bool | float
    Dz: bool | float
    Rx: bool | float
    Ry: bool | float
    Rz: bool | float

class ISupport(BaseModel):
    """Interface for supports"""
    node: str
    supports: ISupportType

# Loads Interface /////////////////////////////////////////////////////////////////////////////////
class INodeLoadType(BaseModel):
    """Interface for node load types"""
    Fx: float
    Fy: float
    Fz: float
    Mx: float
    My: float
    Mz: float

class INodeLoads(BaseModel):
    """Interface for point loads"""
    name: str
    node: str
    loads: INodeLoadType

class IBarPtLoadType(BaseModel):
    """Interface for bar load types"""
    Fx: float
    Fy: float
    Fz: float
    Mx: float
    My: float
    Mz: float

class IBarPointLoads(BaseModel):
    """Interface for bar point loads"""
    name: str
    bar: str
    position: float
    system: Literal['local', 'global']
    loads: IBarPtLoadType

class IBarDistLoadType(BaseModel):
    """Interface for bar load types"""
    Fx: tuple[float, float] # [start, end]
    Fy: tuple[float, float] # [start, end]
    Fz: tuple[float, float] # [start, end]
    Mx: tuple[float, float] # [start, end]
    My: tuple[float, float] # [start, end]
    Mz: tuple[float, float] # [start, end]

class IBarDistLoads(BaseModel):
    """Interface for bar distributed loads"""
    name: str
    bar: str
    position: tuple[float, float]  # [start, end]
    system: Literal['local', 'global']
    loads: IBarDistLoadType

class IBarLoads(BaseModel):
    """Interface for bar loads"""
    point: list[IBarPointLoads]
    distributed: list[IBarDistLoads]

class ILoad(BaseModel):
    """Interface for loads"""
    name: str
    nodes: list[INodeLoads]
    bars: IBarLoads

# Structure Interface /////////////////////////////////////////////////////////////////////////////
class IStructure(BaseModel):
    """Interface"""
    materials: list[IMaterial]
    sections: list[ISection]
    nodes: list[INode]
    bars: list[IBar]
    supports: list[ISupport]
    loads: list[ILoad]
